9223 trial tutorial code

avandeursen opened the pull request:
Fix for TRAC ticket [9223](https://twistedmatrix.com/trac/ticket/9223).

Makes the code used in the [trial tutorial](http://twistedmatrix.com/documents/17.5.0/core/howto/trial.html) compatible with Python 3 (lazy map, floor division, multiple inheritance) and the latest version of Twisted (transfering bytes rather than strings).

avandeursen committed:
Ensure base_3 test_error cases pass

This fix makes the four test_error{Add,Multiply,Divide,Subtract}
test cases in test_base_3.py pass.

base_3.py relied on non-lazy evaluation of the map function.
But in python3 map is lazy, and returns without actually applying the map.

Consequently, a

try:
   map(f, l)
except:
   ...

will never catch exceptions raised by f.

Resolved by moving the try clause into the function.

avandeursen committed:
Ensure Calculus Division operator returns an int

This change makes division tests for base_2 and base_3 pass.

avandeursen committed:
Remove object as RemoteCalculationClient superclass

Inheriting from three classes including object leads to conflicts.

avandeursen committed:
Ensure client and remote exchange bytes, not strings

avandeursen committed:
Add news fragment for update of trial tutorial listings

rodrigc commented on docs/core/howto/listings/trial/calculus/base_3.py:
```diff
@@ -1,16 +1,20 @@
+            raise TypeError("Couldn't coerce argument to integer: %s" % arg)
+
+    def _make_ints(self, *args):
+        return map(self._robust_int, args)
```
Please replace this with a list comprehension, for consistency on Python 2 and 3:

```
return [int(arg) for arg in args]
```
Also, there is no need for the `_robust_int` function.  It is not adding any value, since the existing `ValueError` exception is fine.

rodrigc commented on docs/core/howto/listings/trial/calculus/client_1.py:
```diff
@@ -18,7 +18,7 @@ def lineReceived(self, line):
     def _sendOperation(self, op, a, b):
         d = defer.Deferred()
         self.results.append(d)
-        line = "%s %d %d" % (op, a, b)
```
Since you are modernizing the code, please replace all % modifiers with `.format()` in your pull request.
So this will be:

```
"{} {} {}".format(op, a, b)
```

rodrigc commented on docs/core/howto/listings/trial/calculus/client_1.py:
```diff
@@ -18,7 +18,7 @@ def lineReceived(self, line):
         d = defer.Deferred()
         self.results.append(d)
-        line = "%s %d %d" % (op, a, b)
+        line = ("%s %d %d" % (op, a, b)).encode('utf-8')
```
Additionally, for consistent behavior on Python 2 and 3, any place you are doing an encode on a string, please prefix with *u*, so this will be:

```
u"{} {} {}".format(op, a, b).encode('utf-8')
```

avandeursen commented:
OK, thanks, I will take care of this shortly.

rodrigc commented on docs/core/howto/listings/trial/calculus/base_2.py:
```diff
@@ -11,4 +11,4 @@ def multiply(self, a, b):
 
     def divide(self, a, b):
-        return a / b
+        return a // b
```
One more thing.  Since you are doing division here, please add at the top of the file:

```from __future__ import division```

See: https://docs.python.org/2/library/__future__.html

avandeursen committed:
Revert "Ensure base_3 test_error cases pass"

This reverts commit e7a1f62669c4d56f77486f7e0984ed3771e6fe9c.

Instead of moving the try-except out of the map, we will
replace the map by a list comprehension.

avandeursen committed:
Replace (lazy) map by list comprehension.

This fix makes the four test_error{Add,Multiply,Divide,Subtract}
test cases in test_base_3.py pass.

avandeursen committed:
Replace %-operator by .format(..) function

avandeursen committed:
Prefix strings to be encoded with u

As per request from rodrigc

> for consistent behavior on Python 2 and 3, any place you
> are doing an encode on a string, please prefix with u

avandeursen committed:
Indicate via __future__ that we use floor division

avandeursen commented:
Thanks for the suggestions.

I took care of all, except removing `TypeError`, though I agree it is technically superfluous.

The tutorial in `docs/core/howto/trial.rst` introduces this exception to illustrate testing exceptions and flushing logged errors. So removing it from the code would also entail adjusting the tutorial text itself, as well as the test code.

I'm not opposed to this, but I thought it would be better to leave the tutorial itself the same as much as as possible.

avandeursen committed:
Fix pep8 formating issues on changed code.

rodrigc commented:
Ah, yes, I see `TypeError` is used extensively in the tutorial.  Good point.

rodrigc reviewed (approved):
Good job!  Thanks for going through the tutorial and cleaning things up.  That helps newcomers.

rodrigc committed:
Merge branch 'trunk' into 9223-trial-tutorial-code

rodrigc committed:
Merge branch 'trunk' into 9223-trial-tutorial-code

avandeursen commented:
Thanks! Bumped into this since I am such a newcomer :-)
